deps/chakrashim/src/v8trycatch.cc

```diff
@@ -78,23 +78,34 @@ void TryCatch::GetAndClearException() {
   }
 
   if (hasException) {
-    JsValueRef exceptionRef;
-    errorCode = JsGetAndClearException(&exceptionRef);
+    JsValueRef metadataRef;
+    errorCode = JsGetAndClearExceptionWithMetadata(&metadataRef);
     // We came here through JsHasException, so script shouldn't be in disabled
     // state.
     CHAKRA_ASSERT(errorCode != JsErrorInDisabledState);
     if (errorCode == JsNoError) {
-      error = exceptionRef;
+      metadata = metadataRef;
     }
   }
 }
 
 Handle<Value> TryCatch::ReThrow() {
-  if (error == JS_INVALID_REFERENCE) {
+  if (metadata == JS_INVALID_REFERENCE) {
     GetAndClearException();
   }
 
-  if (error == JS_INVALID_REFERENCE) {
+  if (metadata == JS_INVALID_REFERENCE) {
+    return Local<Value>();
+  }
+
+  JsPropertyIdRef err = jsrt::IsolateShim::GetCurrent()
+    ->GetCachedPropertyIdRef(jsrt::CachedPropertyIdRef::exception);
+  if (err == JS_INVALID_REFERENCE) {
+    return Local<Value>();
+  }
+
+  JsValueRef error;
+  if (JsGetProperty(metadata, err, &error) != JsNoError) {
     return Local<Value>();
   }
```

Comment:
Same code is in 3 functions ReThrow, Exception, StackTrace. Can we refactor and have it in one private function and call that in all these functions?